File: tester/tester_cgi.js
#!/usr/bin/env node
const fs = require('fs');

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function findFilename() {
    const query = process.env.QUERY_STRING || "";
    const target = "file=";
    const foundPos = query.indexOf(target);
    
    if (foundPos === -1) {
        return null;
    }
    
    const startIdx = foundPos + target.length;
    let sepPos = query.indexOf('&', foundPos);
    
    console.error(`${sepPos} ${foundPos}`); // Writes to stderr
    
    let filename = "";
    if (sepPos !== -1) {
        filename = query.substring(startIdx, sepPos);
    } else {
        filename = query.substring(startIdx);
    }
    
    console.error(filename);
    return filename;
}

function handleGet() {
    const filename = findFilename();
    if (!filename) return { status: 404, body: "" };

    try {
        fs.accessSync(filename, fs.constants.F_OK | fs.constants.R_OK);
        const body = fs.readFileSync(filename, 'utf-8');
        return { status: 0, body: body };
    } catch (e) {
        return { status: 404, body: "" };
    }
}

function handlePost(serverBody) {
    const filename = findFilename();
    if (!filename) return { status: 404, body: "" };

    const pathInfo = process.env.PATH_INFO;
    if (pathInfo !== undefined) {
        const filePath = "." + pathInfo;
        try {
            process.chdir(filePath);
        } catch (e) {
            return { status: -1, body: "" };
        }
    }

    console.error(filename);
    
    try {
        fs.writeFileSync(filename, serverBody);
    } catch (e) {
        console.error("Not open");
    }

    return { status: 0, body: "Hi from the script!!!" };
}

async function main() {
    await sleep(5000);

    // Read body from STDIN based on CONTENT_LENGTH
    let serverBody = "";
    const contentLength = parseInt(process.env.CONTENT_LENGTH || "0", 10);
    
    if (contentLength > 0) {
        const buffer = Buffer.alloc(contentLength);
        try {
            let bytesRead = 0;
            while (bytesRead < contentLength) {
                const r = fs.readSync(0, buffer, bytesRead, contentLength - bytesRead, null);
                if (r === 0) break;
                bytesRead += r;
            }
            serverBody = buffer.toString('utf-8', 0, bytesRead);
        } catch (e) {
            // Ignore if reading fails, identical to the C++ silent return
        }
    }

    const method = process.env.REQUEST_METHOD;
    let result = { status: -1, body: "" };

    if (method === "GET") {
        result = handleGet();
    } else if (method === "POST") {
        result = handlePost(serverBody);
    }

    if (result.status !== 0) {
        process.stdout.write("Status: 404 Not Found\r\n");
    }
    
    // Buffer.byteLength ensures correct counting for unicode characters
    const bodyLen = Buffer.byteLength(result.body);
    process.stdout.write(`Content-Length: ${bodyLen}\r\n\r\n`);
    process.stdout.write(result.body);
}

main();

File: tester/tester_cgi.py
#!/usr/bin/env python3
import os
import sys
import time

def read_body():
    """Reads the body from STDIN based on CONTENT_LENGTH."""
    content_length_str = os.environ.get("CONTENT_LENGTH")
    if not content_length_str:
        return ""
    try:
        bytes_2_read = int(content_length_str)
        if bytes_2_read > 0:
            return sys.stdin.read(bytes_2_read)
    except ValueError:
        pass
    return ""

def find_filename():
    """Extracts the 'file=' parameter from QUERY_STRING."""
    query = os.environ.get("QUERY_STRING", "")
    target = "file="
    found_pos = query.find(target)
    
    if found_pos == -1:
        return None
        
    start_idx = found_pos + len(target)
    sep_pos = query.find('&', found_pos)
    
    sys.stderr.write(f"{sep_pos} {found_pos}\n")
    
    if sep_pos != -1:
        filename = query[start_idx:sep_pos]
    else:
        filename = query[start_idx:]
        
    sys.stderr.write(f"{filename}\n")
    return filename

def handle_get():
    filename = find_filename()
    if not filename:
        return 404, ""
        
    if not os.access(filename, os.F_OK | os.R_OK):
        return 404, ""
        
    try:
        with open(filename, "r") as f:
            body = f.read()
        return 0, body
    except IOError:
        return 404, ""

def handle_post(server_body):
    filename = find_filename()
    if not filename:
        return 404, ""
        
    path_info = os.environ.get("PATH_INFO")
    if path_info is not None:
        file_path = "." + path_info
        try:
            os.chdir(file_path)
        except OSError:
            return -1, ""
            
    sys.stderr.write(f"{filename}\n")
    try:
        with open(filename, "w") as f:
            f.write(server_body)
    except IOError:
        sys.stderr.write("Not open\n")
        
    return 0, "Hi from the script!!!"

def main():
    time.sleep(5)
    
    server_body = read_body()
    method = os.environ.get("REQUEST_METHOD", "")
    
    status = 0
    body = ""
    
    if method == "GET":
        status, body = handle_get()
    elif method == "POST":
        status, body = handle_post(server_body)
    else:
        status = -1
        
    if status != 0:
        sys.stdout.write("Status: 404 Not Found\r\n")
        
    sys.stdout.write(f"Content-Length: {len(body)}\r\n\r\n")
    sys.stdout.write(body)

if __name__ == "__main__":
    main()
